Better registry

chriskrycho opened the pull request:
[Rendered](https://github.com/chriskrycho/ember-rfcs/blob/better-registry/text/0000-improved-ember-registry-apis.md)

chriskrycho committed:
Registry APIs: create RFC from template

chriskrycho committed:
Registry APIs: add the MOAR STRINGS alternative

chriskrycho committed:
Registry APIs: add summary statement

chriskrycho committed:
Registry APIs: add an unresolved question

chriskrycho committed:
Registry APIs: add some alternatives

chriskrycho committed:
Registry APIs: add (incomplete) proposed types

chriskrycho committed:
Registry APIs: add a (very secondary) motivation

chriskrycho committed:
Registry APIs: correct mistaken assertion about alternative

chriskrycho committed:
Registry APIs: expand example of confusion

chriskrycho committed:
Registry APIs: fill out motivation section

chriskrycho committed:
Registry APIs: clarify additional registry

chriskrycho committed:
Registry APIs: include example in summary

chriskrycho committed:
Registry APIs: improve TS motivation wording

chriskrycho committed:
Registry APIs: move most TS motivation to appendix

chriskrycho committed:
Registry APIs: fix internal link to TS appendix

chriskrycho committed:
Registry APIs: provide overview of detailed design

chriskrycho committed:
Registry APIs: describe Resolver `schemaVersion`

chriskrycho committed:
Registry APIs: include an additional motivation

chriskrycho committed:
Registry APIs: define and explain `Identifier`

chriskrycho committed:
Registry APIs: clarify a sentence about `Identifier`

chriskrycho committed:
Registry APIs: add 'Do Nothing' alternative

chriskrycho committed:
Registry APIs: rename Resolver section for consistency

chriskrycho committed:
Registry APIs: note additional uses for `schemaVersion`

chriskrycho committed:
Registry APIs: stub out sections for all Owner APIs

chriskrycho committed:
Registry APIs: define new API for `Owner.lookup`

chriskrycho committed:
Registry APIs: define new API for `Owner.inject`

chriskrycho committed:
Registry APIs: use "method" instead of "function" where appropriate

chriskrycho committed:
Registry APIs: explain `LookupOptions` type

chriskrycho committed:
Registry APIs: fix heading level of "Owner APIs"

chriskrycho committed:
Registry APIs: define new API for `Owner.factoryFor`

chriskrycho committed:
Registry APIs: expand proposed type definitions

chriskrycho committed:
Registry APIs: remove unused `Class` helper type

chriskrycho committed:
Registry APIs: define new API for `Owner.hasRegistration`

chriskrycho committed:
Registry APIs: clean up a few Owner API sections

- Fix a typo: `a strings` -> `strings`.
- Remove `LookupOptions` where it's no longer needed courtesy of the
  addition made in 1b82df24.

chriskrycho committed:
Registry APIs: define new API for `Owner.register`

chriskrycho committed:
Registry APIs: define new API for `Owner.registerOptions`

chriskrycho committed:
Registry APIs: consolidate definition of `RegisterOptions`

chriskrycho committed:
Registry APIs: relocate `FactoryTypeIdentifier` discussion

chriskrycho committed:
Registry APIs: define new API for `Owner.registerOptionsForType`

chriskrycho committed:
Registry APIs: move note about TS APIs to Detailed Design

chriskrycho committed:
Registry APIs: define new API for `Owner.registeredOption`

chriskrycho committed:
Registry APIs: stub out discussion of codemod

chriskrycho committed:
Registry APIs: placeholder for deprecation message

chriskrycho committed:
Registry APIs: define new API for `Object.registeredOptions`

chriskrycho committed:
Registry APIs: specify lookup by equality in Owner overview

chriskrycho committed:
Registry APIs: define new API for `Owner.registeredOptionsForType`

chriskrycho committed:
Registry APIs: define new API for `Owner.resolveRegistration`

This API is probably not *quite* right, but I'll wait for discussion
on the RFC to iterate on it.

chriskrycho committed:
Registry APIs: clarify Factory and FactoryManager

chriskrycho committed:
Registry APIs: define new API for `Owner.unregister`

chriskrycho committed:
Registry APIs: simplify detailed design section

- Convert all the differences to a single diff, since the text was
  effectively duplicated across them.
- Explain the gist of the changes at the start of the section.

chriskrycho committed:
Registry APIs: add deprecation message

chriskrycho committed:
Registry APIs: add drawbacks

chriskrycho committed:
Registry APIs: improve wording in Alternatives

chriskrycho committed:
Registry APIs: explain "How we teach this"

chriskrycho committed:
Registry APIs: supply working TS type definitions

chriskrycho committed:
Registry APIs: correct part of TS detailed motivation

chriskrycho committed:
Registry APIs: clean up one type definition

chriskrycho committed:
Registry APIs: update start date assuming next Monday

chriskrycho committed:
Registry APIs: fix one alternative write-up

chriskrycho committed:
Registry APIs: extend unanswered questions section

chriskrycho committed:
Registry APIs: improve teaching guide first bullet

chriskrycho committed:
Registry APIs: further type definitions: `Class` type

chriskrycho committed:
Registry APIs: fix another type design error

cafreeman commented on text/0000-improved-ember-registry-apis.md:
```diff
@@ -0,0 +1,540 @@
+
+As is the case in the microsyntax-based design, these factory type injections may not be namespaced.
+
+### `Resolver`
```
If you've got prior art regarding the use of versioned Ember internals, I think it would be good to link to them in this section and call this out as an already accepted practice within the framework

chriskrycho replied:
That's a helpful note – thank you! I did so in d1b5a73, and added some more explicit language on backwards compatibility there at the same time.

rwjblue replied:
https://github.com/emberjs/ember.js/pull/9994 is a good example IMHO, and it is *exactly* related to the resolver capabilities (and backwards compat guarding, much the same as this RFC)

cafreeman commented on text/0000-improved-ember-registry-apis.md:
```diff
@@ -0,0 +1,540 @@
+
+> You invoked `Owner.lookup` with a `string` full name: `'<type>:<name>'`. This usage is deprecated and will be removed in Ember 4.0. Instead, pass an object identifier: `{ type: '<type>', name: '<name>' }`.
+
+Here, the message should substittue the *actual* passed type and name. For example, if the user tried to look up a service named `session` by invoking `lookup('service:session')` the deprecation will read:
```
typo on `substitute`

chriskrycho replied:
Thank you! Pushing a fix for it now.

chriskrycho committed:
Registry APIs: add notes on compat, schema

chriskrycho committed:
Registry APIs: fix a typo in Deprecations section

rwjblue commented on text/0000-improved-ember-registry-apis.md:
```diff
@@ -0,0 +1,544 @@
+
+Here the `type` corresponds to the prefix component of the legacy API, and `name` to the postfix component of the legacy API: `service:foo` becomes `{ type: 'service', name: 'foo' }`.
+
+The `namespace` field allows for identifiers to be restricted to a particular namespace, as in for example tools which namespace lookups for addons (e.g. [ember-holy-futuristic-template-namespacing-batman](https://github.com/rwjblue/ember-holy-futuristic-template-namespacing-batman)). It is optional since it is not required in normal usage.
```
It is also useful to point out that the current ember-resolver _already_ supports this functionality (without experimental addons). For example:

```js
export default class MyService extends Service {
  @service('addon-name@service-name') blah;
}
```

chriskrycho replied:
Ah, excellent point: I will add that note.

chriskrycho replied:
Captured in 0d6b6687.

cafreeman commented on text/0000-improved-ember-registry-apis.md:
```diff
@@ -0,0 +1,540 @@
+
+This is a less flexible proposal, and—more importantly—it has a significant downside for usability beyond the simplest case. Supporting namespaces here would require *another* string argument, and functions with multiple arguments of the same type are notoriously easy to misuse. (“Wait, what’s the order? Namespace first or type first? Does namespace come after type, or after name?” etc.)
+
+### String-based API as “sugar”
```
I question whether this is even worth mentioning as a possible additional API given the problems outlined in the previous "New string based API" section. Given that this RFC's general thrust seems to be "trading away sugar for explicitness and unlocking better tooling possibilities", it almost feels weird to even put the option of keeping some of the sugar around into the conversation.

chriskrycho replied:
I broadly agree, but figured I'd raise it explicitly because it *will* come up in the RFC conversation proper. Better to call it out, I think?

chriskrycho replied:
I broadly agree, but figured I'd raise it explicitly because it *will* come up in the RFC conversation proper. Better to call it out, I think?

rwjblue replied:
Ya, agree that it seems good to include. But I ultimately hope we decide not to implement it this way...

chriskrycho replied:
Concur! I added some notes on why the RFC recommends against this, hopefully capturing both of your points; see b4b07a6.

chriskrycho committed:
Registry APIs: explain rollout support and timeline

rwjblue commented on text/0000-improved-ember-registry-apis.md:
```diff
@@ -0,0 +1,544 @@
+
+### `Resolver`
+
+Ember’s `Resolver` function is public API, designed to be customized and overridden. To support backwards compatibility with existing custom resolvers, we introduce a `schemaVersion` key to the `Resolver` API, which may be `undefined` or an integer value. If `schemaVersion` is `undefined` or `0`, the legacy behavior of the resolver is maintained:
```
We should consider following suit with modifiers and component managers which use a `capabilities` property for this.

https://github.com/emberjs/rfcs/blob/master/text/0373-Element-Modifier-Managers.md#capabilities
https://github.com/emberjs/rfcs/blob/master/text/0213-custom-components.md#capabilities

chriskrycho replied:
I like this idea; I'll rewrite the `schemaVersion` in terms of a `capabilities` and move `schemaVersion` to the **Alternatives** section!

→ 52676e5

cafreeman commented on text/0000-improved-ember-registry-apis.md:
````diff
@@ -0,0 +1,540 @@
++ }
+```
+
+This *works*, but is additional boilerplate and, more importantly, *duplication of information*, which can get out of sync. Both of these problems are avoided by the new design proposed by this RFC. With the new design, users would still need the same boilerplate required today, but the types for `lookup` and other registry APIs can be extended in a way transparent to users so that they “just wor ...
````
>  With the new design, users would still need the same boilerplate required today, but the types for `lookup` and other registry APIs can be extended in a way transparent to users so that they “just work” with *zero* additional effort, as covered in the next section.

It might be worth calling this out at the beginning of the TypeScript section. I came into this section with the primary question "does this mean we don't have to do those module declarations anymore??" and I imagine I won't be the only one. Making it clear from the get-go what you will (and won't) gain in TypeScript-land as a result of this change might help avoid confusion for the less than careful reader.

chriskrycho replied:
That's a great point! "Not making it *worse*" is the story; I'll clarify as much. Done in e569ff76.

chriskrycho committed:
Registry APIs: mention existing namespacing microsyntax

rwjblue commented on text/0000-improved-ember-registry-apis.md:
```diff
@@ -0,0 +1,544 @@
+
+The other registry API changes may also be implemented in terms of the new resolver `schemaVersion` check.
+
+This design is completely backwards compatible: it continues working exactly as it does today, with users opting *into* the change by upgrading their version of `ember-resolver`, which will implement the new design proposed by this RFC as a breaking change.
```
I think this is conceptually fine (making the upgrade of ember-resolver a breaking change update), but I think it would actually be fine to do compatibly.

Something like changing [this line](https://github.com/ember-cli/ember-resolver/blob/8f624f294eb4a2c77bb40e619d7cf347f63ecd17/addon/resolvers/classic/index.js#L150) to:

```js
resolve(fullNameOrIdentifier) {
  let parsedName = typeof fullNameOrIdentifier === 'string' ? this.parseStringName(fullNameOrIdentifier) : this.parseIdentifier(fullNameOrIdentifier);
  // ...snip...
}
```

chriskrycho replied:
Interesting! I think I *prefer* the breaking change just for the clarity, but perhaps we'd be better to do the same thing here as in general: introduce in a compatible way, but deprecate, and then drop in a later major?

chriskrycho replied:
Incorporated the suggestions here in e0efbc56!

rwjblue commented on text/0000-improved-ember-registry-apis.md:
```diff
@@ -0,0 +1,556 @@
+
+### Owner APIs
+
+The owner APIs all change to use `Identifier` or `FactoryTypeIdentifier` instead of strings. 
```
```suggestion
The owner APIs all change to use `Identifier` (or `FactoryTypeIdentifier` as appropriate) instead of strings. 
```

chriskrycho committed:
Registry APIs: clarify intent & impact on TS users

rwjblue commented on text/0000-improved-ember-registry-apis.md:
```diff
@@ -0,0 +1,556 @@
+
+#### `Factory` and `FactoryManager`
+
+Some APIs refer to *factories* and *factory managers*. The API for these classes is the same as it was previously, with the exception that the `fullName` and `normalizedNames` are deprecated:
```
We should replace these APIs with _something_ (it is hard to deprecate them otherwise). Maybe:

```ts
interface Factory<T, C extends FactoryClass | object = FactoryClass> {
  class?: C;
  identifier: FactoryIdentifier,
  fullName?: string; // DEPRECATED
  normalizedName?: string; // DEPRECATED
  create(props?: { [prop: string]: any }): T;
}
```

chriskrycho replied:
Ah, thank you—I hadn't thought of that. 😅 

→ 64310896

chriskrycho committed:
Registry APIs: clarify introductory sentence in Owner APIs

Co-Authored-By: Robert Jackson <me@rwjblue.com>

rwjblue commented on text/0000-improved-ember-registry-apis.md:
```diff
@@ -0,0 +1,556 @@
+
+### Codemod
+
+All existing microsyntax invocations can be straightforwardly migrated to the new syntax with a codemod.
```
I think this is largely true, but not 100%. Specifically cases where the string is DRY'ed up and an imported variable. In those cases the codemod can't disambiguate between that argument being a string or a POJO.

I don't think its super common, but seems good to make it clear that the codemod can convert 100% of "static string" arguments to these APIs, and _some_ dynamism (but not all).

chriskrycho replied:
Ah, excellent clarification. I'll update the text here to capture that more clearly.

→ 242be4a

rwjblue commented on text/0000-improved-ember-registry-apis.md:
```diff
@@ -0,0 +1,556 @@
+
+All existing microsyntax invocations can be straightforwardly migrated to the new syntax with a codemod.
+
+### Deprecation messaging
```
We have generally been including the actual prose for the deprecation guide in deprecation RFC's lately. I don't think it necessarily has to be done before submitting, but we'd want it before landing.

chriskrycho replied:
Ah, I didn't realize that; I will happily add it!

→ 32a2c3f

chriskrycho committed:
Registry APIs: handle `ember-resolver` backwards-compatibly

rwjblue commented on text/0000-improved-ember-registry-apis.md:
```diff
@@ -0,0 +1,556 @@
+
+- We will supply a polyfill for the new behavior, supporting *at least* the current LTS. The behavior here can be polyfilled to many if not all existing LTS releases; we would likely limit this slightly, but would support *many* LTS releases.
+
+Changes to `ember-resolver` will be introduced in a major version, signaling clearly to users that there is a breaking change inbound.
```
See note above, I don't actually think this needs a breaking change.

chriskrycho replied:
Whoops, meant to update this one as well when I updated the other notes about back-compat! Thank you!

→ 923f508

rwjblue commented on text/0000-improved-ember-registry-apis.md:
```diff
@@ -0,0 +1,556 @@
+
+- **New concepts:** this RFC introduces the ideas of `Identifier` and `FactoryTypeIdentifier` to define the new identifier types passed as arguments to the various registry functions. The basic shapes of these objects, and perhaps their names (as inline text like “identifier” and “factory type identifier”) will need to be integrated into the guides and API docs. See also [**Unresolved Questions** ...
+
+- The existing documentation about dependency injection can largely remain unchanged—mostly, it can be simplified in certain key areas. Paragraphs explaining the microsyntax can be eliminated or simplified by showing the identifier object type. New Ember users will continue to benefit from the detailed introduction to the concepts in the guides, but will have one fewer concept to learn along the  ...
```
in the first bit, should probably remove either `largely` or `mostly` (using both reads "funny")

chriskrycho replied:
Heh, very good catch. Pushing a fix now!

→ resolved in 458a0b5

rwjblue commented on text/0000-improved-ember-registry-apis.md:
```diff
@@ -0,0 +1,556 @@
+
+- The existing APIs have served Ember well for a long time. Introducing any new change involves some risk, particularly in areas so core to how Ember behaves.
+
+- The proposed API here trades away brevity in favor of clarity, tooling support, and being “normal” JavaScript. `'service:session'` is many fewer characters to type than `{ type: 'service', name: 'session' }`.
```
I know I'm not supposed to argue with the drawbacks, but ffs when will people realize that we should be optimizing for _reading_ (which we do massively more) not _writing_ (which we only do once for any given line) :sob:

chriskrycho replied:
You and I *very strongly agree* here. I will actually add a note to that effect: I recognize people have the feeling, but thing it's *substantially* outweighed by the reading and maintenance tradeoffs.

→ c79945c

rwjblue commented on text/0000-improved-ember-registry-apis.md:
````diff
@@ -0,0 +1,556 @@
+lookup({ service: 'foo' })
+```
+
+While this is nice from the perspective of the consumer of the API, it muddies the API substantially and makes implementation more complex and with much worse performance: the lookup would have to iterate over all keys on the object passed every time, mapping those keys to the *known* identifer interface and then checking the others against the registry. This also makes the TypeScript implementat ...
````
What would it mean to do `lookup({ service: 'foo', route: 'bar' })`!?!?

chriskrycho replied:
Oh my, I hadn't even thought of that as a downside. 😂 That's *terrible*.

→ def1a9f

rwjblue commented on text/0000-improved-ember-registry-apis.md:
```diff
@@ -0,0 +1,556 @@
+
+## Unresolved questions
+
+- What is the right name for the parameter and interface for the lookup? `identifier` is a reasonable choice, but does overlap with the notion of identifiers from Ember Data. It is also *quite* generic. We could use the longer but perhaps clearer `RegistrationIdentifier` or `RegistryIdentifier` or perhaps `RegistryId` for the type when discussing it.
```
I'll throw `FactoryIdentifier` in the ring...

chriskrycho replied:
I could be persuaded that `FactoryIdentifier` and `FactoryTypeIdentifier` make good sense here! Added a couple more options including that one.

→ 3e12eaf

chriskrycho committed:
Registry APIs: improve wording around teaching

chriskrycho committed:
Registry APIs: elaborate tradeoffs in one drawback

chriskrycho committed:
Registry APIs: note another drawback of object-based alternative

chriskrycho committed:
Registry APIs: more alternatives to `(FactoryType)?Identifier`

chriskrycho committed:
Registry APIs: update a missed note about ember-resolve back-compat

chriskrycho committed:
Registry APIs: clarify limits of a codemod

chriskrycho committed:
Registry APIs: clarify intent of diff views

chriskrycho committed:
Registry APIs: supply a replacement API for names on Factory(Manager)

chriskrycho committed:
Registry APIs: note drawbacks of string-based 'sugar' API

chriskrycho committed:
Registry APIs: add deprecation text

chriskrycho committed:
Registry APIs: redesign Resolver in terms of capabilities

chriskrycho committed:
Registry APIs: add missing TypeRegistry to type defs

The `TypeRegistry` did not previously exist, so should be documented,
and adding it clarifies the need to define registries for routes and
components... *somewhere*. Components in particular will require some
careful consideration, since there is no single natural home for them
as there is for many of the other registries: a single codebase may
have registrations for both Ember and Glimmer components, for example.

chriskrycho committed:
Registry APIs: add table of contents

chriskrycho committed:
Registry APIs: improve motivation headings

dfreeman commented on text/0000-improved-ember-registry-apis.md:
````diff
@@ -0,0 +1,719 @@
+```diff
+ interface Registry {
+-  session: Session;
++  session: typeof Session;
````
Do you have any concerns about the global-ness of `Registry` types and how that plays with the isolated registries/containers that engines have? For engines developed in a separate repo from each other/their host app, the possibility of cross-contamination is probably pretty low, but in a monorepo or in-app-engine setup, it seems like projects are likely to wind up with the union of all the available services/controllers/etc.

I realize this is already the status quo, but since this RFC touches on and expands the use of type registries, it seemed like it might be worth raising, even if the answer is ultimately still "it's probably fine."

chriskrycho replied:
My take is roughly: It's *not* really fine, and we need to develop a solution for it—it's on my roadmap for work I expect to be doing in the last three quarters of this year 🤞—but (a) the ideas introduced here don't meaningfully *worsen* the situation and (b) the solution we eventually build to solve it *should* work just as well.

I have some still-very-unformed ideas around ways we could leverage the changes proposed in this RFC, particularly around the use of namespaces. We might, for example, be able to tweak the registries to include the namespace for addons (including engines), and then (a) recommend to TS users that they consistently use the namespace rather than letting it be inferred or (b) see if we can figure out some way to infer based on context what the namespace should be [probably impossible?] or (c) do something that combines (a) with TS' project configurations to get the right outcomes. Like I said: *very* un-formed as of yet, but the ideas become much more tenable if this RFC ends up accepted and the API is rationalized accordingly.

dfreeman commented on text/0000-improved-ember-registry-apis.md:
```diff
@@ -0,0 +1,719 @@
+
+type InstanceOf<T> = T extends new (...args: any) => infer R ? R : never;
+
+type ConstructorFor<T> = new (...args: ConstructorArgs<T>) => InstanceOf<T>;
```
I'm probably missing something subtle (or even completely obvious), but what does this type do? At first glance it looks like it would be a no-op if applied to a type with a `new` signature, other than possibly throwing out any potential parametric polymorphism.

chriskrycho replied:
…huh. Let me play with this. Because now that you mention it, yeah, it should be. 😰

Seems like as long as the *other* constraints are in place in the spots where we use this, and we deploy `typeof` appropriately in the registry, we shouldn't need this, because we'd already have it.

chriskrycho replied:
I validated that; thank you for catching it. I'll update it later with a change that drops this entirely and makes the type param for `FactoryManager` and `Factory` be `Class` (instead of `T`), perhaps constrained to be something which extends `new (...args: any[]) => any`.

→ 4227463

dfreeman commented on text/0000-improved-ember-registry-apis.md:
````diff
@@ -0,0 +1,719 @@
+})
+```
+
+The Typed Ember blueprints generate that boilerplate for users, which (hopefully) minimizes the pain of working with these definitions. However, to make this work for the registry APIs, users (and therefore the blueprints) would have to add *another* declaration to the bottom of the file:
````
This is _well_ outside the scope of this RFC, but one of the things I'd love to explore (likely within the realm of that `@typeset` project we've discussed) is automating the maintenance of these type registries instead of relying on boilerplate merged from the source modules.

chriskrycho replied:
I had that point very much in mind myself as I was writing this. If we can programmatically generate/maintain those, that would make users' lives *much* easier. I experimented with that a bit in my previous gig, but never got to something I like.

chriskrycho committed:
Registry APIs: clarify a sentence about microsyntax

chriskrycho committed:
Registry APIs: add service and controller APIs

chriskrycho committed:
Registry APIs: eliminate extraneous type

chriskrycho committed:
Registry APIs: types for session and controller

chriskrycho committed:
Registry APIs: mention service and controller in summary

chriskrycho committed:
Registry APIs: include examples for service, controller in summary
